packages/pkl.lua/lua.pkl

```diff
@@ -0,0 +1,292 @@
+  /// inline style, although any nested collections may be rendered in the multiline style.
+  ///
+  /// Note that [Map] and [List] are always rendered in the inline style.
+  multilineThreshold: Int = 2
+
+  /// Value converters to apply before values are rendered.
+  ///
+  /// For more information see [ValueRenderer.converters]. Note that due to language limitations,
+  /// any entries in a [Dynamic] will have converters applied as though the entry was a property.
+  converters: Mapping<(Class|String), (unknown) -> Any>
+
+  extension = "lua"
+
+  function renderValue(value: Any): String =
+    let (path = List("^"))
+    render(convert(value, path), path, 0)
+
+  function renderDocument(value: Any): String =
+    let (path = List("^"))
+    let (value = convert(value, path))
+    if (value is RenderDirective) "\(value.text)\n"
+    else if (value is LuaRenderDirective) "\(value.before ?? "")\(render(value.value, path, 0))\(value.after ?? "")\n"
+    else if (value is Dynamic|Typed)
+      value.toMap().fold("", (acc, k, v) ->
+        let (path = path.add(Prop(k)))
+        let (v = convert(v, path))
+        if (omitNullProperties && v == null && k is String) acc else
+        acc + "\(renderKey(k, "_ENV")) = \(render(v, path, 0))\n")
+      + (if (value is Dynamic) value.toList() else List()).foldIndexed("", (idx, acc, v) ->
+        let (path = path.add(splatKey))
+        let (v = convert(v, path))
+        // remember, Lua indexes are 1-based
+        acc + "_ENV[\(idx+1)] = \(render(v, path, 0))\n")
+    else if (value is Mapping|Map)
+      value.fold("", (acc, k, v) ->
+        let (path = path.add(Key(k)))
+        let (v = convert(v, path))
+        acc + "\(renderKey(k, "_ENV")) = \(render(v, path, 0))\n")
+    else throw("The top-level value of a Lua document must have type `Typed`, `Dynamic`, `Mapping`, or `Map`, but got type `\(value.getClass())`")
+
+  // ## Misc helpers ##
+
+  // Returns a [Prop] for string keys, otherwise [splatKey].
+  local function Prop(k: Any): Prop|Key =
+    if (k is String) new Prop { name = k } else splatKey
+
+  // Returns a [Key] for string keys, otherwise [splatKey]
+  local function Key(k: Any): Key =
+    if (k is String) new Key { key = k } else splatKey
+
+  // ## Converters ##
+
+  local converterMap = converters.toMap()
+
+  // path specs are already in reversed order
+  local pathConverters: List<Pair<List<PathEntry>,(unknown) -> Any>> = converterMap
+  .filter((key, _) -> key is String)
+  .mapKeys((key, _) ->
+    if (key.matches(pathSpecRegex))
+      if (key == "") List("^") else // PcfRenderer treats empty path as "^"
+      key.split(pathSpecSplitRegex)
+      .map((it) ->
+        if (it == "^") it
+        else if (it.startsWith("[")) new Key { key = it.drop(1).dropLast(1) }
+        else new Prop { name = it })
+      .reverse()
+    else throw("Converter path `\(key)` has invalid syntax."))
+  .entries
```

Comment:
```suggestion
  // noinspection TypeMismatch
  local pathConverters: List<Pair<List<PathEntry>,(unknown) -> Any>> =
    converterMap
      .filter((key, _) -> key is String)
      .mapKeys((key, _) ->
        if (key.matches(pathSpecRegex))
          if (key == "") List("^")
          else // PcfRenderer treats empty path as "^"
            key
              .split(pathSpecSplitRegex)
              .map((it) ->
                if (it == "^") it
                else if (it.startsWith("[")) new Key { key = it.drop(1).dropLast(1) }
                else new Prop { name = it })
              .reverse()
        else throw("Converter path `\(key)` has invalid syntax."))
      .entries
```

Reply:
What does `// noinspection TypeMismatch` do?

Reply:
Nevermind, I just checked out the IntelliJ plugin and now I understand.